cli/data.go

```diff
@@ -0,0 +1,194 @@
+			},
+			IncludeBinary: true,
+			CountOnly:     false,
+		})
+		// TODO: Make sure EOF is properly interpreted. Iirc rpc errors aren't properly parsed by errors.Is.
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		mds := resp.GetMetadata()
+		if len(mds) != 1 {
+			return errors.Errorf("expected a single response, received %d", len(mds))
+		}
+		data := resp.GetData()
+		if len(data) != 1 {
+			return errors.Errorf("expected a single response, received %d", len(data))
+		}
+
+		md := mds[0]
+		datum := data[0]
+		mdJsonBytes, err := protojson.Marshal(md)
+		if err != nil {
+			return err
+		}
+		jsonFile, err := os.Create(filepath.Join(dst, "metadata", datum.GetId()+".json"))
+		if err != nil {
+			return err
+		}
+		if _, err := jsonFile.Write(mdJsonBytes); err != nil {
+			return err
+		}
+
+		// TODO: Include file name in metadata.
+
+		gzippedBytes := datum.GetBinary()
+		r, err := gzip.NewReader(bytes.NewBuffer(gzippedBytes))
+		if err != nil {
+			return err
+		}
+
+		// TODO: We need to store file extension too. In sync we map from ext -> mime type, so this is already available.
+		//       Or maybe we can just get this from file name.
+		ext := mimeTypeToFileExt(md.GetMimeType())
+
+		dataFile, err := os.Create(filepath.Join(dst, "data", datum.GetId()+ext))
+		if _, err := io.Copy(dataFile, r); err != nil {
+			return err
+		}
+		if err := r.Close(); err != nil {
+			return err
+		}
+		skip++
+	}
+
+	return nil
+}
+
+func (c *AppClient) TabularData(dst string, filter *datapb.Filter) error {
+	if err := c.ensureLoggedIn(); err != nil {
+		return err
+	}
+
+	// Make dst/data and dst/metadata directory.
+	if err := os.MkdirAll(filepath.Join(dst, dataDir), os.ModePerm); err != nil {
+		return err
+	}
+	if err := os.MkdirAll(filepath.Join(dst, metadataDir), os.ModePerm); err != nil {
+		return err
+	}
+
+	resp, err := c.dataClient.TabularDataByFilter(context.Background(), &datapb.TabularDataByFilterRequest{
+		DataRequest: &datapb.DataRequest{
+			Filter: filter,
+			// TODO: For now don't worry about skip/limit. Just do everything in one request. Can implement batching when
+			//       tabular is implemented.
+		},
+		CountOnly: false,
+	})
+	if err != nil {
+		return err
+	}
+	mds := resp.GetMetadata()
+
+	for i, md := range mds {
+		mdJsonBytes, err := protojson.Marshal(md)
+		if err != nil {
+			return errors.Wrap(err, "error marshaling metadata")
+		}
+		mdFile, err := os.Create(filepath.Join(dst, "metadata", strconv.Itoa(i)+".json"))
+		if err != nil {
+			return errors.Wrapf(err, "error creating metadata file %s", mdFile.Name())
+		}
+		if _, err := mdFile.Write(mdJsonBytes); err != nil {
+			return errors.Wrapf(err, "error writing metadata file %s", mdFile.Name())
+		}
+		if err := mdFile.Close(); err != nil {
+			return errors.Wrapf(err, "error closing metadata file %s", mdFile.Name())
+		}
+	}
+
+	data := resp.GetData()
+	// TODO: Use textpb insted of ndjson, and save multiple files.
```

Comment:
I think we should have a CLI flag of output type (json, protobuf, csv, etc) that is consistent between how we're writing out the metadata and data files. If we start with only JSON, we can add that flag with a follow-up PR that potentially supports another type (e.g. .textpb files).

Reply:
Agreed. Want me to make a ticket for that?

Reply:
That'd be great, thanks! The ticket would include making the decision of which formats to support.

Reply:
Made [this](https://viam.atlassian.net/browse/DATA-640), will update with a link to the CLI when this is merged